rust/operator-binary/src/spark_k8s_controller.rs

```diff
@@ -239,41 +306,106 @@ pub async fn reconcile(spark_application: Arc<SparkApplication>, ctx: Arc<Ctx>)
     Ok(Action::await_change())
 }
 
+fn init_containers(
+    spark_application: &SparkApplication,
+    logging: &Logging<SparkContainer>,
+) -> Result<Vec<Container>> {
+    let mut jcb = ContainerBuilder::new(&SparkContainer::Job.to_string())
+        .context(IllegalContainerNameSnafu)?;
+    let job_container = spark_application.spec.image.as_ref().map(|job_image| {
+        let mut args = Vec::new();
+        if let Some(ContainerLogConfig {
+            choice: Some(ContainerLogConfigChoice::Automatic(log_config)),
+        }) = logging.containers.get(&SparkContainer::Job)
+        {
+            args.push(capture_shell_output(
+                VOLUME_MOUNT_PATH_LOG,
+                &SparkContainer::Job.to_string(),
+                log_config,
+            ));
+        };
+        args.push(format!("echo Copying job files to {VOLUME_MOUNT_PATH_JOB}"));
+        args.push(format!("cp /jobs/* {VOLUME_MOUNT_PATH_JOB}"));
+        // Wait until the log file is written.
+        args.push("sleep 1".into());
+
+        jcb.image(job_image)
+            .command(vec!["/bin/bash".to_string(), "-c".to_string()])
+            .args(vec![args.join(" && ")])
+            .add_volume_mount(VOLUME_MOUNT_NAME_JOB, VOLUME_MOUNT_PATH_JOB)
+            .add_volume_mount(VOLUME_MOUNT_NAME_LOG, VOLUME_MOUNT_PATH_LOG)
+            .build()
+    });
+
+    let spark_image = spark_application
+        .spec
+        .spark_image
+        .as_deref()
+        .context(ObjectHasNoSparkImageSnafu)?;
+
+    let mut rcb = ContainerBuilder::new(&SparkContainer::Requirements.to_string())
+        .context(IllegalContainerNameSnafu)?;
+    let requirements_container = spark_application.requirements().map(|req| {
+        let mut args = Vec::new();
+        if let Some(ContainerLogConfig {
+            choice: Some(ContainerLogConfigChoice::Automatic(log_config)),
+        }) = logging.containers.get(&SparkContainer::Requirements)
+        {
+            args.push(capture_shell_output(
+                VOLUME_MOUNT_PATH_LOG,
+                &SparkContainer::Requirements.to_string(),
+                log_config,
+            ));
+        };
+        args.push(format!(
+            "echo Installing requirements to {VOLUME_MOUNT_PATH_REQ}: {req}"
+        ));
+        args.push(format!(
+            "pip install --target={VOLUME_MOUNT_PATH_REQ} {req}"
+        ));
```

Comment:
Was in before already. This would fail in an offline cluster? @razvan @adwk67 i created an issue https://github.com/stackabletech/spark-k8s-operator/issues/227

Reply:
Thanks!